File: ChatCommon/ByteStringConverter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CharCommon
{
    public static class ByteStringConverter
    {
        private const string Z85Alphabet = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-:+=^!/*?&<>()[]{}@%$#";

        private static readonly int[] s_decodeMap = CreateDecodeMap();

        public static string ToZ85String(byte[] data)
        {
            if (data is null) throw new ArgumentNullException(nameof(data));

            if (data.Length == 0)
                return "0|";

            string z85 = EncodeZ85(data);
            return data.Length + "|" + z85;
        }

        public static byte[] FromZ85String(string s)
        {
            if (s is null) throw new ArgumentNullException(nameof(s));
            if (s.Length == 0) return Array.Empty<byte>();

            var sepIndex = s.IndexOf('|');
            if (sepIndex <= 0) throw new FormatException("Invalid format: missing length prefix.");

            if (!int.TryParse(s.Substring(0, sepIndex), out var length) || length < 0)
                throw new FormatException("Invalid length prefix.");

            var payload = s.Substring(sepIndex + 1);
            if (length == 0) return Array.Empty<byte>();

            var decoded = DecodeZ85(payload);
            if (decoded.Length < length)
                throw new FormatException("Decoded data shorter than expected length.");

            if (decoded.Length == length) return decoded;
            var result = new byte[length];
            Array.Copy(decoded, 0, result, 0, length);
            return result;
        }

        private static string EncodeZ85(byte[] data)
        {
            int pad = (4 - (data.Length % 4)) % 4;
            int total = data.Length + pad;
            var dst = new StringBuilder((total / 4) * 5);

            var buf = new byte[total];
            Array.Copy(data, buf, data.Length);

            for (int i = 0; i < total; i += 4)
            {
                uint value = ((uint)buf[i] << 24) | ((uint)buf[i + 1] << 16) | ((uint)buf[i + 2] << 8) | buf[i + 3];
                var block = new char[5];
                for (int j = 4; j >= 0; j--)
                {
                    block[j] = Z85Alphabet[(int)(value % 85)];
                    value /= 85;
                }
                dst.Append(block);
            }

            return dst.ToString();
        }

        private static byte[] DecodeZ85(string s)
        {
            if (string.IsNullOrEmpty(s)) return Array.Empty<byte>();
            if (s.Length % 5 != 0) throw new FormatException("Invalid Z85 payload length.");

            var outBytes = new List<byte>(s.Length / 5 * 4);

            for (int i = 0; i < s.Length; i += 5)
            {
                ulong value = 0;
                for (int j = 0; j < 5; j++)
                {
                    char c = s[i + j];
                    int v = c < 256 ? s_decodeMap[c] : -1;
                    if (v < 0) throw new FormatException($"Invalid character '{c}' in Z85 payload.");
                    value = value * 85 + (uint)v;
                }

                outBytes.Add((byte)((value >> 24) & 0xFF));
                outBytes.Add((byte)((value >> 16) & 0xFF));
                outBytes.Add((byte)((value >> 8) & 0xFF));
                outBytes.Add((byte)(value & 0xFF));
            }

            return outBytes.ToArray();
        }

        private static int[] CreateDecodeMap()
        {
            var map = Enumerable.Repeat(-1, 256).ToArray();
            for (int i = 0; i < Z85Alphabet.Length; i++)
            {
                map[(int)Z85Alphabet[i]] = i;
            }
            return map;
        }
    }
}

File: ChatClientAndroid/app/src/main/java/com/simplechat/android/ByteStringConverter.java
package com.simplechat.android;

public class ByteStringConverter {
    private static final char[] Z85_ENCODER =
            "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-:+=^!/*?&<>()[]{}@%$#".toCharArray();
    private static final int[] Z85_DECODER = new int[256];
    static {
        for (int i = 0; i < Z85_DECODER.length; i++) Z85_DECODER[i] = -1;
        for (int i = 0; i < Z85_ENCODER.length; i++) Z85_DECODER[Z85_ENCODER[i]] = i;
    }

    // Encodes a byte array to a Z85 string
    public static String toZ85String(byte[] data) {
        if (data == null) throw new IllegalArgumentException("data is null");
        if (data.length == 0) return "0|";
        String z85 = encodeZ85(data);
        return data.length + "|" + z85;
    }

    // Decodes a Z85 string to a byte array
    public static byte[] fromZ85String(String s) {
        if (s == null) throw new IllegalArgumentException("input is null");
        if (s.length() == 0) return new byte[0];
        int sepIndex = s.indexOf('|');
        if (sepIndex <= 0) throw new IllegalArgumentException("Invalid format: missing length prefix.");
        int length = Integer.parseInt(s.substring(0, sepIndex));
        if (length < 0) throw new IllegalArgumentException("Invalid length prefix.");
        String payload = s.substring(sepIndex + 1);
        if (length == 0) return new byte[0];
        byte[] decoded = decodeZ85(payload);
        if (decoded.length < length) throw new IllegalArgumentException("Decoded data shorter than expected length.");
        if (decoded.length == length) return decoded;
        byte[] result = new byte[length];
        System.arraycopy(decoded, 0, result, 0, length);
        return result;
    }

    private static String encodeZ85(byte[] data) {
        int pad = (4 - (data.length % 4)) % 4;
        int total = data.length + pad;
        StringBuilder sb = new StringBuilder((total / 4) * 5);
        byte[] buf = new byte[total];
        System.arraycopy(data, 0, buf, 0, data.length);
        for (int i = 0; i < total; i += 4) {
            long value = ((long)buf[i] & 0xFF) << 24 |
                         ((long)buf[i + 1] & 0xFF) << 16 |
                         ((long)buf[i + 2] & 0xFF) << 8 |
                         ((long)buf[i + 3] & 0xFF);
            char[] block = new char[5];
            for (int j = 4; j >= 0; j--) {
                block[j] = Z85_ENCODER[(int)(value % 85)];
                value /= 85;
            }
            sb.append(block);
        }
        return sb.toString();
    }

    private static byte[] decodeZ85(String s) {
        if (s == null || s.isEmpty()) return new byte[0];
        if (s.length() % 5 != 0) throw new IllegalArgumentException("Invalid Z85 payload length.");
        byte[] outBytes = new byte[s.length() / 5 * 4];
        int outIndex = 0;
        for (int i = 0; i < s.length(); i += 5) {
            long value = 0;
            for (int j = 0; j < 5; j++) {
                char c = s.charAt(i + j);
                int v = c < 256 ? Z85_DECODER[c] : -1;
                if (v < 0) throw new IllegalArgumentException("Invalid character '" + c + "' in Z85 payload.");
                value = value * 85 + v;
            }
            outBytes[outIndex++] = (byte)((value >> 24) & 0xFF);
            outBytes[outIndex++] = (byte)((value >> 16) & 0xFF);
            outBytes[outIndex++] = (byte)((value >> 8) & 0xFF);
            outBytes[outIndex++] = (byte)(value & 0xFF);
        }
        return outBytes;
    }
}
